### ISSUE TITLE
[SCM-868] gitexe add() does not return added files when invoked in subdir

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-scm#1097]
**[Ilya Basin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=basinilya)** opened **[SCM-868](https://issues.apache.org/jira/browse/SCM-868?redirect=false)** and commented

I'm going to add a wagon-scm test suite for git. One of the test cases is calling the GitAddCommand command with:

```java
repo: file:///C:/progs/maven/maven-wagon/wagon-providers/wagon-scm/target/test-classes/test-repo-git-test/
fileSet: basedir = C:\Users\basin\AppData\Local\Temp\wagon-scm2020146470.checkout\file-list; files = [test-resource.txt]
```

After adding the files the command internally calls "git rev-parse --show-toplevel" which prints:

```java
C:/Users/basin/AppData/Local/Temp/wagon-scm2020146470.checkout
```

And on the next line it tries to do (pseudo code):

```java
("file:/C:/Users/basin/AppData/Local/Temp/wagon-scm1144102340.checkout/file-list/").relativize(URI.create(
"C:/Users/basin/AppData/Local/Temp/wagon-scm2020146470.checkout"))
```

This is so wrong! What were those people from SCM-709 trying to get? A double dot ".."? The argument to the relativize() method MUST be a child, not a parent, otherwise the argument is returned unchanged.

There are so many reasons why getting an absolute path from git is a bad idea: Symlinks, Windows paths.

Nevertheless, there's a simple solution. The original problem was that "git status --porcelain" printed paths relative to the top level instead of relative to the base dir.

So we should just call

```
git rev-parse --show-prefix
```

instead of "show-toplevel" and strip this prefix from the paths printed by "git status"

---

**Affects:** 1.9.5, 1.10.0

**Issue Links:**
- [SCM-709](https://issues.apache.org/jira/browse/SCM-709) REGRESSION: git status doesn't work if repository root is not the working directory

**Remote Links:**
- [GitHub Pull Request #66
](https://github.com/apache/maven-scm/pull/66)
- [GitHub Pull Request #66
](https://github.com/apache/maven-scm/pull/66)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Its hard for me to understand the given issue and your proposed solution because I am completely unacquiant with the code.

**Comment by jira-importer** [comment]
**[Ilya Basin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=basinilya)** commented

If you run the "add" or "checkin" command on Windows with the current working directory and the env. var. "TMP" being on different drive letters, then the command result will be "zero files added" or "zero files checked in".

Unfortunately, the unit tests of maven-scm don't check this. This issue, however, breaks WagonTestCase.testWagon() in maven-wagon when attempted to run for gitexe.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Can you please rebase your changes against master as well as provide a test project where I can see the faulty behavior? I am still having problem to understand the root cause, but am willing to merge your work.

**Comment by jira-importer** [comment]
**[Ilya Basin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=basinilya)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) Done.

I modified GitCheckInCommandNoBranchTest to demonstrate the problem. The system property "scm.test.tmpdir" now alters the test directory location. Default is "target/".

Please checkout my branch, do "mvn install -Dmaven.test.skip", change dir to 

```
maven-scm-providers\maven-scm-providers-git\maven-scm-provider-gitexe
```

and run:

```
mvn test -Dtest="GitCheckInCommandNoBranchTest#testCheckinNoBranch" -Dscm.test.tmpdir=D:\
```

It should succeed. Then revert the last commit (my fix) and repeat. Given that your source tree is not on "D:\", It should fail with:

```java
GitCheckInCommandNoBranchTest.testCheckinNoBranch:76 expected:<1> but was:<0>
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [3b0f2ecb900366564f6ea735302819c3a2c6513f](https://gitbox.apache.org/repos/asf?p=maven-scm.git;a=commit;h=3b0f2ecb900366564f6ea735302819c3a2c6513f).
